src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1524,47 +1591,91 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 }
             }
             Reservation(WriteKind::MutableBorrow(BorrowKind::Mut { .. }))
-            | Write(WriteKind::MutableBorrow(BorrowKind::Mut { .. })) => if let Err(place_err) =
-                self.is_mutable(place, is_local_mutation_allowed)
-            {
-                error_reported = true;
-
-                let item_msg = match self.describe_place(place) {
-                    Some(name) => format!("immutable item `{}`", name),
-                    None => "immutable item".to_owned(),
-                };
-
-                let mut err = self.tcx
-                    .cannot_borrow_path_as_mutable(span, &item_msg, Origin::Mir);
-                err.span_label(span, "cannot borrow as mutable");
+            | Write(WriteKind::MutableBorrow(BorrowKind::Mut { .. })) => {
+                match self.is_mutable(place, is_local_mutation_allowed) {
+                    Ok(Place::Local(local))
```

Comment:
Nit: I find this setup confusing. Can we do this instead?

```rust
Ok(Place::Local(local)) => {
    if is_local_mutation_allowed != ... { ... }
}
```

and then remove the *other* `Place::Local` arm below? 

(In general, I prefer to avoid `if` clauses on match arms except in particular cases, e.g. with very trivial bodies or where it enables a lot more re-use.)